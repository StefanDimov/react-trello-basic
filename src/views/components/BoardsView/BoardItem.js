import React from 'react'
import * as boardUtils from '../../../utils/boardUtils'
import { DropdownButton, MenuItem } from 'react-bootstrap'

const DELETE_EVENT_KEY = 1

/**
 * Component that represents a Board Item. Ment to be used with BoardItemsList Component
 * @class comp/BoardItem
 * @param {object} board - The Board data
 * @param {function} onDeleteBoard - function to execute when the board should be deleted
 */
export default class BoardItem extends React.Component {

    constructor(props) { // eslint-disable-line
        super(props)

        this.preventEventPropagation = this.preventEventPropagation.bind(this)
        this.onActionSelect = this.onActionSelect.bind(this)
    }

    /**
     * Stops event propagation
     * @param {object} e the event object
     */
    preventEventPropagation(e) {
        e.stopPropagation()
    }

    /**
     * Handler for the dropdown actions
     * @param {number} eventKey the eventKey number of the MenuItem
     * @param {object} e the event obejct
     */
    onActionSelect(eventKey, e) {
        e.stopPropagation()

        switch (eventKey) {
        case DELETE_EVENT_KEY: // Delete option
            this.props.onDeleteBoard(this.props.board)
            break
        }
    }

    render() { // eslint-disable-line
        const { board } = this.props
        return (
            <div className="my-board-item">
                <div className="my-board-item-title">{ board.title }</div>
                <DropdownButton
                    onClick={this.preventEventPropagation}
                    onSelect={this.onActionSelect}
                    pullRight
                    className="my-board-item-dropdown-overrides"
                    bsStyle="link"
                    title=""
                    id={board.id + '-dropdown'}>
                        <MenuItem eventKey={DELETE_EVENT_KEY}>Delete</MenuItem>
                </DropdownButton>
            </div>
        )
    }
}

BoardItem.propTypes = {
    board: boardUtils.boardPropType.isRequired,
    onDeleteBoard: React.PropTypes.func.isRequired
}
