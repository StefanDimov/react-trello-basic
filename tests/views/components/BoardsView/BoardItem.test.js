import React from 'react'
import { mount } from 'enzyme'
import renderer from 'react-test-renderer'

import { getEmptyBoard } from '../../../_mocks/Board.mocks.js'

import BoardItem from '../../../../src/views/components/BoardsView/BoardItem'

describe('BoardItem', () => {

    let board
    const onDeleteBoard = jest.fn()

    beforeEach(() => {
        board = getEmptyBoard()
        onDeleteBoard.mockReset()
    })

    describe('rendering', () => {
        it('should render correctly', () => {
            const tree = renderer.create(
                <BoardItem board={board} onDeleteBoard={onDeleteBoard} />
            ).toJSON()

            expect(tree).toMatchSnapshot()
        })
    })

    describe('dropdown options', () => {
        it('should properly call onDeleteBoard callback on Delete option click', () => {
            const wrapper = mount(<BoardItem board={board} onDeleteBoard={onDeleteBoard} />)

            wrapper.find('.dropdown button').simulate('click')
            wrapper.find('.dropdown li').first().find('a').simulate('click')

            expect(onDeleteBoard).toHaveBeenCalledTimes(1)
            expect(onDeleteBoard).toHaveBeenCalledWith(board)
        })
    })
})
