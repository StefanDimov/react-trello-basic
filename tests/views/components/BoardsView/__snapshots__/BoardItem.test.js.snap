exports[`BoardItem rendering should render correctly 1`] = `
<div
  className="my-board-item">
  <div
    className="my-board-item-title">
    boardTitle
  </div>
  <div
    className="dropdown btn-group btn-group-link">
    <button
      aria-expanded={false}
      aria-haspopup={true}
      className="my-board-item-dropdown-overrides dropdown-toggle btn btn-link"
      disabled={false}
      id="boardId-dropdown"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      type="button">
       
      <span
        className="caret" />
    </button>
    <ul
      aria-labelledby="boardId-dropdown"
      className="dropdown-menu dropdown-menu-right"
      role="menu">
      <li
        className=""
        role="presentation"
        style={undefined}>
        <a
          href="#"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="menuitem"
          tabIndex="-1">
          Delete
        </a>
      </li>
    </ul>
  </div>
</div>
`;
